editoast/src/views/authz.rs

```diff
@@ -0,0 +1,229 @@
+        .map_err(AuthzError::from)?
+    {
+        return Err(NoSuchUserError::NoSuchUser { user_id }.into());
+    }
+
+    Ok(Json(Roles {
+        builtin: authorizer
+            .user_builtin_roles(user_id)
+            .await
+            .map_err(AuthzError::from)?
+            .into_iter()
+            .map(|b| b.as_str().to_owned())
+            .collect(),
+        application: authorizer
+            .infer_application_roles(user_id)
+            .await
+            .map_err(AuthzError::from)?,
+    }))
+}
+
+#[derive(serde::Deserialize, utoipa::ToSchema)]
+struct RoleListBody {
+    roles: Vec<String>,
+}
+
+#[derive(Debug, thiserror::Error, EditoastError)]
+#[editoast_error(base_id = "authz:role")]
+enum InvalidRoleTag {
+    #[error("Invalid role tag: {0}")]
+    #[editoast_error(status = 400)]
+    Invalid(String),
+}
+
+#[utoipa::path(
+    post, path = "",
+    tag = "authz",
+    params(UserIdPathParam),
+    request_body = inline(RoleListBody),
+    responses(
+        (status = 204, description = "The roles have been granted sucessfully"),
+    ),
+)]
+async fn grant_roles(
+    Path(UserIdPathParam { user_id }): Path<UserIdPathParam>,
+    Extension(mut authorizer): AuthorizerExt,
+    Json(RoleListBody { roles }): Json<RoleListBody>,
+) -> Result<impl axum::response::IntoResponse> {
+    if !authorizer
+        .check_roles([BuiltinRole::SubjectRead, BuiltinRole::RoleWrite].into())
+        .await
+        .map_err(AuthorizationError::from)?
+    {
+        return Err(AuthorizationError::Unauthorized.into());
+    }
+
+    if !authorizer
+        .user_exists(user_id)
+        .await
+        .map_err(AuthzError::from)?
+    {
+        return Err(NoSuchUserError::NoSuchUser { user_id }.into());
+    }
+
+    let roles = authorizer
+        .roles_config
+        .resolve(roles.iter().map(String::as_str))
+        .map_err(|unknown| InvalidRoleTag::Invalid(unknown.to_owned()))?;
+    authorizer
+        .grant_roles(user_id, roles)
+        .await
+        .map_err(AuthzError::from)?;
+    Ok(axum::http::StatusCode::NO_CONTENT)
+}
+
+#[utoipa::path(
+    delete, path = "",
+    tag = "authz",
+    params(UserIdPathParam),
+    request_body = inline(RoleListBody),
+    responses(
+        (status = 204, description = "The roles have been removed sucessfully"),
+    ),
+)]
+async fn strip_roles(
+    Path(UserIdPathParam { user_id }): Path<UserIdPathParam>,
+    Extension(mut authorizer): AuthorizerExt,
+    Json(RoleListBody { roles }): Json<RoleListBody>,
+) -> Result<impl axum::response::IntoResponse> {
+    if !authorizer
+        .check_roles([BuiltinRole::SubjectRead, BuiltinRole::RoleWrite].into())
+        .await
+        .map_err(AuthorizationError::from)?
+    {
+        return Err(AuthorizationError::Unauthorized.into());
+    }
+
+    if !authorizer
+        .user_exists(user_id)
+        .await
+        .map_err(AuthzError::from)?
+    {
+        return Err(NoSuchUserError::NoSuchUser { user_id }.into());
+    }
```

Comment:
I factorized the `user_exists` part but not the `check_roles` header to maintain consistency with the other handlers (at least until we have the middleware). I let you resolve if it's ok.

Reply:
I would really factorize the entirety of it because these 3 functions will actually always require the same code and we do want that, no? My point being if we change it at some point in the future, we might forget to change the 2 others and I feel there is a higher chance that we do indeed want to update all 3 everytime, what do you think?

Reply:
Alright, done in [`01fe017` (#8984)](https://github.com/OpenRailAssociation/osrd/pull/8984/commits/01fe0173f5317adea09f4564805f231ef67f9c1a)

Reply:
As discussed, reverted since the required roles slightly differ in these endpoints.